Workiva/over_react_codemod
# CPLAT-12621 Codemod to update factory syntax for Dart 2.9

## Motivation
  
We need a codemod to update factory syntax to the new Dart 2.9 boilerplate.

Factory Declarations:
```diff
- UiFactory<FooProps> Foo = _$Foo; // ignore: undefined_identifier, invalid_assignment
+ UiFactory<FooProps> Foo = castUiFactory(_$Foo); // ignore: undefined_identifier
```

Function Components:
```diff
UiFactory<FooProps> Foo = uiFunction(
    (props) {},
-   $FooConfig, // ignore: undefined_identifier, argument_type_not_assignable
+   _$FooConfig, // ignore: undefined_identifier
);
```

Connected Components:
```diff
UiFactory<FooProps> Foo =
    connect<SomeState, FooProps>(
  mapStateToProps: (state) => (Foo()
    ..foo = state.foo
    ..bar = state.bar
  ),
- )(_$Foo); // ignore: undefined_identifier, argument_type_not_assignable
+ )(castUiFactory(_$Foo); // ignore: undefined_identifier
```

## Changes
  
- Created `dart2_9_upgrade` codemod executable to bump the `over_react` lower bound and run the following suggestors:
  - `FactoryAndConfigIgnoreCommentRemover`: to remove ignore comments (particularly `argument_type_not_assignable` and `invalid_assignment`).
  - `FactoryConfigMigrator`: to add the private prefix to generated factory configs.
  - `GeneratedFactoryMigrator`: to wrap generated factories in the `castUiFactory` function.
- Wrote tests.

#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        
      - [ ] Verify that tests cover all cases
      - [ ] Pull down this branch and:
        - [ ] Run `pub global activate -s path .` in the codemod repo
        - [ ] Then run `pub global run over_react_codemod:dart2_9_upgrade` in multiple repos and verify that it works as expected.
          - [ ] Verify that all connected components, component factory declarations, function components (and `uiForwardRef`) usages are updated correctly.
          - [ ] Verify that codemod runs without errors even when parsing invalid code (specifically in the `over_react` analyzer plugin playground).
          - [ ] Verify that the codemod makes no additional changes the second time it is run.
          - [ ] Verify that factory declarations in the legacy boilerplate syntax are not updated.
          - [ ] Verify that `argument_type_not_assignable` and `invalid_assignment` ignore comments are removed (and other ignore comments are not)
            - To more thoroughly QA `FactoryAndConfigIgnoreCommentRemover`, try adding `FactoryAndConfigIgnoreCommentRemover('undefined_identifier')` [here](https://github.com/Workiva/over_react_codemod/blob/e7c6ae082db31d02f9094c1cf6f961d64ade5ad8/lib/src/executables/dart2_9_upgrade.dart#L55-L56) and verify that all `undefined_identifier` ignore comments are removed (regardless of their location around the generated factory).

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#manual-testing-criteria
